tests/slowproxy_test.go

```diff
@@ -48,40 +54,74 @@ func (s *slowProxy) Serve() error {
 		if err != nil {
 			return err
 		}
-		// Run handleConnection async since server dial might block
-		go s.handleConnection(conn)
+		// Run handleConnection async since it blocks while its connection is open, collect errors.
+		go func() {
+			if err := s.handleConnection(conn); err != nil {
+				s.muConnErrors.Lock()
+				s.connErrors = append(s.connErrors, fmt.Errorf("handleConnection: %w", err))
+				s.muConnErrors.Unlock()
+			}
+		}()
 	}
 }
 
+// ConnectionErrors returns a slice of errors from closed connections since the last time Errors was called.
+func (s *slowProxy) ConnectionErrors() []error {
+	var errors []error
+	s.muConnErrors.Lock()
+	copy(errors, s.connErrors)
+	s.connErrors = nil
+	s.muConnErrors.Unlock()
+	return errors
+}
+
 // Close the proxy server and frees network resources. Already running connections are not stopped.
 func (s *slowProxy) Close() error {
 	return s.listener.Close()
 }
 
 // handleConnection sets up and handles a single proxy connection.
-func (s *slowProxy) handleConnection(client_conn net.Conn) error {
+func (s *slowProxy) handleConnection(clientConn net.Conn) (err error) {
+	defer errors.Capture(&err, clientConn.Close)()
+
 	// Open up connection to destination server
-	server_conn, err := net.Dial("tcp", s.dest)
+	serverConn, err := net.Dial("tcp", s.dest)
 	if err != nil {
-		client_conn.Close()
-		return nil
+		return
+	}
+	defer errors.Capture(&err, serverConn.Close)()
+
+	slowServer := limiter.NewWriter(serverConn, s.bytesPerSecond, s.burstLimit)
+	slowClient := limiter.NewWriter(clientConn, s.bytesPerSecond, s.burstLimit)
+
+	errorChan := make(chan error, 2)
+	var wg sync.WaitGroup
+	shuttleData := func(dst io.WriteCloser, src io.ReadCloser) {
+		defer wg.Done()
+		errorChan <- func(dst io.WriteCloser, src io.ReadCloser) (err error) {
+			defer errors.Capture(&err, dst.Close)()
+			defer errors.Capture(&err, src.Close)()
+			_, err = io.Copy(dst, src)
+			return
+		}(dst, src)
 	}
 
-	slow_server := limiter.NewWriter(server_conn, s.bytesPerSecond, s.burstLimit)
-	slow_client := limiter.NewWriter(client_conn, s.bytesPerSecond, s.burstLimit)
-
-	go s.shuttleData(slow_server, client_conn)
-	go s.shuttleData(slow_client, server_conn)
-
-	return nil
-}
+	wg.Add(2)
+	go shuttleData(slowServer, clientConn)
+	go shuttleData(slowClient, serverConn)
+	wg.Wait()
+	close(errorChan)
+
+	for shuttleErr := range errorChan {
+		// io.EOF is an expected condition, not an error for us
+		if shuttleErr != nil && shuttleErr != io.EOF {
```

Comment:
You could use `errors.Is()` here

Reply:
I'll leave this for now. `errors.Is()` is definitely preferable, although for backwards compatibility reasons Go guarantees it'll return an unwrapped `io.EOF` when you try to read past the end of a file / stream / socket. When there was a proposal to update the standard libraries to use `errors.Is()`, it was declined. See https://github.com/golang/go/issues/39155/